lib/scheduler.ml

```diff
@@ -428,23 +430,44 @@ module Make (Streamd : StreamDescriptor) = struct
           (* We check for activity again, because the stream may have gone
            * inactive after the call to `write` above. *)
           written, Streamd.requires_output descriptor
-        else (
-          match PriorityQueue.pop p.children with
-          | Some ((id, (Stream i as i_node)), children') ->
-            p.t_last <- i.t;
-            let written, subtree_is_active = schedule i_node in
-            if subtree_is_active
-            then (
-              update_t i_node written;
-              p.children <- PriorityQueue.add id i_node children')
-            else (
-              implicitly_close_idle_stream i.descriptor max_seen_ids;
-              p.children <- children');
-            written, subtree_is_active
-          | None ->
-            (* Queue is empty, see line 6 above. *)
-            0, false)
+        else traverse p_node children
+    and traverse : type a. a node -> PriorityQueue.t -> int * bool =
+     fun p_node ->
+      let rec loop children =
+        match PriorityQueue.pop children with
+        | Some ((id, (Stream i as i_node)), children') ->
+          let written, subtree_is_active = schedule i_node in
+          if not subtree_is_active
+          then (
+            implicitly_close_idle_stream i.descriptor max_seen_ids;
+            (* XXX(anmonteiro): we may not want to remove from the tree right
+             * away. *)
+            remove_child p_node id);
```

Comment:
I feel like this was already wrong: I don't think we can remove a child without moving all its children to the parent node too.

Reply:
I think #201 fixes this: a subtree is not active if it has children.